### ISSUE TITLE
deadlock in zfs_AcquireForLazyWrite

### ISSUE THREAD

**Issue by EchterAgo** [openzfsonwindows/openzfs#313]
I just encountered a deadlock in `zfs_AcquireForLazyWrite` it seems:

```
Possible deadlock. Use !locks FFFFBD81D41E4250 to determine the resource owner
Break instruction exception - code 80000003 (first chance)
nt!ExpWaitForResource+0x20d770:
fffff800`800249c0 cc              int     3
1: kd> !locks FFFFBD81D41E4250

Resource @ 0xffffbd81d41e4250    Exclusively owned
    Contention Count = 1
    NumberOfExclusiveWaiters = 1
     Threads: ffffbd81a5175080-01<*> 

     Threads Waiting On Exclusive Access:
1 total locks
1: kd> .thread FFFFBD81D41E4250
Invalid thread @ ffffbd81`d41e4250 type - context unchanged.
1: kd> k
 # Child-SP          RetAddr               Call Site
00 ffffd00d`2d45f6e0 fffff800`7fe2114e     nt!ExpWaitForResource+0x20d770
01 ffffd00d`2d45f760 fffff800`8a11d989     nt!ExAcquireResourceExclusiveLite+0x1fe
02 ffffd00d`2d45f7f0 fffff800`7fe579e0     OpenZFS!zfs_AcquireForLazyWrite+0x1f9 [H:\dev\openzfs\module\os\windows\zfs\zfs_vnops_windows.c @ 160] 
03 ffffd00d`2d45f870 fffff800`7ff057b1     nt!CcWriteBehindInternal+0x130
04 ffffd00d`2d45f950 fffff800`7ff02a99     nt!CcWriteBehind+0x91
05 ffffd00d`2d45fa40 fffff800`7fe50545     nt!CcWorkerThread+0x259
06 ffffd00d`2d45fb70 fffff800`7ff0e6f5     nt!ExpWorkerThread+0x105
07 ffffd00d`2d45fc10 fffff800`80006278     nt!PspSystemThreadStartup+0x55
08 ffffd00d`2d45fc60 00000000`00000000     nt!KiStartSystemThread+0x28
1: kd> .thread ffffbd81a5175080
Implicit thread is now ffffbd81`a5175080
1: kd> k
  *** Stack trace for last set context - .thread/.cxr resets it
 # Child-SP          RetAddr               Call Site
00 ffffd00d`2af24800 fffff800`7fe1bca0     nt!KiSwapContext+0x76
01 ffffd00d`2af24940 fffff800`7fe1b1cf     nt!KiSwapThread+0x500
02 ffffd00d`2af249f0 fffff800`7fef7eee     nt!KiCommitThreadWait+0x14f
03 ffffd00d`2af24a90 fffff800`89e3998a     nt!KeWaitForMultipleObjects+0x2be
04 ffffd00d`2af24ba0 fffff800`89f11440     OpenZFS!spl_cv_wait+0xea [H:\dev\openzfs\module\os\windows\spl\spl-condvar.c @ 120] 
05 ffffd00d`2af24c30 fffff800`89f10c59     OpenZFS!txg_wait_synced_impl+0x290 [H:\dev\openzfs\module\zfs\txg.c @ 710] 
06 ffffd00d`2af24cb0 fffff800`8a035e5c     OpenZFS!txg_wait_synced+0x29 [H:\dev\openzfs\module\zfs\txg.c @ 738] 
07 ffffd00d`2af24d10 fffff800`8a03535e     OpenZFS!dmu_tx_wait+0x2ac [H:\dev\openzfs\module\zfs\dmu_tx.c @ 1269] 
08 ffffd00d`2af24d70 fffff800`8a1616ad     OpenZFS!dmu_tx_assign+0x17e [H:\dev\openzfs\module\zfs\dmu_tx.c @ 1207] 
09 ffffd00d`2af24de0 fffff800`8a160ebf     OpenZFS!zfs_extend+0x27d [H:\dev\openzfs\module\os\windows\zfs\zfs_znode.c @ 1627] 
0a ffffd00d`2af24e70 fffff800`8a12b6cc     OpenZFS!zfs_freesp+0x12f [H:\dev\openzfs\module\os\windows\zfs\zfs_znode.c @ 1851] 
0b ffffd00d`2af24fe0 fffff800`8a12c4c4     OpenZFS!zfs_write_wrap+0x88c [H:\dev\openzfs\module\os\windows\zfs\zfs_vnops_windows.c @ 5028] 
0c ffffd00d`2af251c0 fffff800`8a12c9ee     OpenZFS!fs_write_impl+0x364 [H:\dev\openzfs\module\os\windows\zfs\zfs_vnops_windows.c @ 5298] 
0d ffffd00d`2af25290 fffff800`8a132d1b     OpenZFS!fs_write+0x47e [H:\dev\openzfs\module\os\windows\zfs\zfs_vnops_windows.c @ 5420] 
0e ffffd00d`2af25380 fffff800`8a12f132     OpenZFS!fsDispatcher+0x187b [H:\dev\openzfs\module\os\windows\zfs\zfs_vnops_windows.c @ 7216] 
0f ffffd00d`2af254f0 fffff800`7fe10665     OpenZFS!dispatcher+0x292 [H:\dev\openzfs\module\os\windows\zfs\zfs_vnops_windows.c @ 7327] 
10 ffffd00d`2af255e0 fffff800`7e91710f     nt!IofCallDriver+0x55
11 ffffd00d`2af25620 fffff800`7e914a43     FLTMGR!FltpLegacyProcessingAfterPreCallbacksCompleted+0x28f
12 ffffd00d`2af25690 fffff800`7fe10665     FLTMGR!FltpDispatch+0xa3
13 ffffd00d`2af256f0 fffff800`7e914a76     nt!IofCallDriver+0x55
14 ffffd00d`2af25730 fffff800`7fe10665     FLTMGR!FltpDispatch+0xd6
15 ffffd00d`2af25790 fffff800`8020142c     nt!IofCallDriver+0x55
16 ffffd00d`2af257d0 fffff800`801ce0d9     nt!IopSynchronousServiceTail+0x34c
17 ffffd00d`2af25870 fffff800`802687c6     nt!IopWriteFile+0x23d
18 ffffd00d`2af25970 fffff800`80010ef8     nt!NtWriteFile+0x996
19 ffffd00d`2af25a90 00007ffa`3330d0e4     nt!KiSystemServiceCopyEnd+0x28
1a 00000000`2ebffbf8 00000000`00000000     0x00007ffa`3330d0e4
```

**Comment by lundman** [comment]
Not entirely convinced, could simply be it took too long and Windows reported it as a deadlock. 
OpenZFS!txg_wait_synced_impl+0x290  is syncing to disk, waiting for writes to complete,

But, it could be that the txg backend has stalled, and as the first symptom of that, is the deadlock detection kicks in. So I don't think it is a deadlock, but rather txg slow/hung.

**Comment by EchterAgo** [comment]
Rclone did actually stall before the debugger broke. I'll try to let it continue running.

**Comment by EchterAgo** [comment]
I think this is actually related to #283 / kmem too:

[cbuf.txt](https://github.com/openzfsonwindows/openzfs/files/13224236/cbuf.txt)

**Comment by EchterAgo** [comment]
The kernel itself seems to also be trying to allocate:

```
11: kd> k
 # Child-SP          RetAddr               Call Site
00 ffffd00d`2bfd4958 fffff800`7ff4a5a7     nt!memset+0xd3
01 ffffd00d`2bfd4960 fffff800`7feeb015     nt!MiDeleteSystemPageTables+0x4f
02 ffffd00d`2bfd4aa0 fffff800`7fec0fad     nt!MiReturnSystemVa+0x131
03 ffffd00d`2bfd4bc0 fffff800`7fec0db2     nt!MmFreePoolMemory+0x1c9
04 ffffd00d`2bfd4c40 fffff800`7ff15a2c     nt!RtlpHpEnvFreeVA+0x12
05 ffffd00d`2bfd4c70 fffff800`7ff158de     nt!RtlpHpVaMgrRangeFree+0x54
06 ffffd00d`2bfd4ca0 fffff800`7fec0d96     nt!RtlpHpVaMgrCtxFree+0x6a
07 ffffd00d`2bfd4d00 fffff800`7ff898ec     nt!RtlpHpFreeVA+0x8a
08 ffffd00d`2bfd4d40 fffff800`8006edbe     nt!RtlpHpSegMgrRelease+0x60
09 ffffd00d`2bfd4dc0 fffff800`7ff16158     nt!RtlpHpSegMgrAllocate+0x157cba
0a ffffd00d`2bfd4e20 fffff800`7fe9281c     nt!RtlpHpSegSegmentAllocate+0x2c
0b ffffd00d`2bfd4e80 fffff800`7fe9203a     nt!RtlpHpSegPageRangeAllocate+0x69c
0c ffffd00d`2bfd4f20 fffff800`7fe11c56     nt!RtlpHpSegAlloc+0x5a
0d ffffd00d`2bfd4f80 fffff800`805b8074     nt!ExAllocateHeapPool+0x8f6
0e ffffd00d`2bfd50c0 fffff800`84a2f92e     nt!ExAllocatePoolWithTag+0x64
0f ffffd00d`2bfd5110 fffff800`84a1547e     Wof!FileProvReadCompressedOnNewStack+0x32e
10 ffffd00d`2bfd51a0 fffff800`7fe72228     Wof!FileProvReadCompressedCallout+0xe
11 ffffd00d`2bfd51d0 fffff800`7fe7219d     nt!KeExpandKernelStackAndCalloutInternal+0x78
12 ffffd00d`2bfd5240 fffff800`84a31e53     nt!KeExpandKernelStackAndCalloutEx+0x1d
13 ffffd00d`2bfd5280 fffff800`84a31c7e     Wof!FileProvReadCompressedSwitchStacks+0x53
14 ffffd00d`2bfd52c0 fffff800`84a125e0     Wof!FileProvCbRead+0x12e
15 ffffd00d`2bfd5320 fffff800`7e9164cb     Wof!WofPreReadCallback+0x270
16 ffffd00d`2bfd53e0 fffff800`7e915f7a     FLTMGR!FltpPerformPreCallbacksWorker+0x36b
17 ffffd00d`2bfd5500 fffff800`7e915021     FLTMGR!FltpPassThroughInternal+0xca
18 ffffd00d`2bfd5550 fffff800`7e914a2b     FLTMGR!FltpPassThrough+0x541
19 ffffd00d`2bfd55e0 fffff800`7fe10665     FLTMGR!FltpDispatch+0x8b
1a ffffd00d`2bfd5640 fffff800`7e91710f     nt!IofCallDriver+0x55
1b ffffd00d`2bfd5680 fffff800`7e914a43     FLTMGR!FltpLegacyProcessingAfterPreCallbacksCompleted+0x28f
1c ffffd00d`2bfd56f0 fffff800`7fe10665     FLTMGR!FltpDispatch+0xa3
1d ffffd00d`2bfd5750 fffff800`7feea2c7     nt!IofCallDriver+0x55
1e ffffd00d`2bfd5790 fffff800`7fe958aa     nt!IoPageReadEx+0x1d7
1f ffffd00d`2bfd5800 fffff800`7fe94b1d     nt!MiIssueHardFaultIo+0xb6
20 ffffd00d`2bfd5850 fffff800`7fe30798     nt!MiIssueHardFault+0x29d
21 ffffd00d`2bfd5960 fffff800`8000d1d8     nt!MmAccessFault+0x468
22 ffffd00d`2bfd5b00 00007ffa`1e54a3b3     nt!KiPageFault+0x358
23 0000000b`e3a7f780 00000000`00000000     0x00007ffa`1e54a3b3
```

**Comment by lundman** [comment]
yeah put it as lower priority for now, its most likely a symptoms of the other thing

**Comment by EchterAgo** [comment]
Yea, I now think so too, closing. I also tried to let it continue running but it does not seem to recover.

**Comment by EchterAgo** [comment]
`kstat` outputs from before the stall: [kstat_logs.zip](https://github.com/openzfsonwindows/openzfs/files/13224474/kstat_logs.zip)
